sycl/unittests/Extensions/CommandGraph/Update.cpp

```diff
@@ -150,3 +151,255 @@ TEST_F(CommandGraphTest, UpdateRangeErrors) {
+TEST_F(WholeGraphUpdateTest, LessNodes) {
+  // Test that using an update graph that has less nodes results in an error.
+
+  auto NodeA = Graph.add(EmptyKernel);
+  auto NodeB =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeC =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeD =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeB, NodeC));
+
+  auto UpdateNodeA = UpdateGraph.add(EmptyKernel);
+  auto UpdateNodeB = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  auto UpdateNodeC = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  // NodeD is missing in the update
+
+  auto GraphExec = Graph.finalize(experimental::property::graph::updatable{});
+  EXPECT_THROW(GraphExec.update(UpdateGraph), sycl::exception);
+}
+
+TEST_F(WholeGraphUpdateTest, ExtraEdges) {
+  // Test that using an update graph with extra nodes results in an error.
+
+  auto NodeA = Graph.add(EmptyKernel);
+  auto NodeB =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeC =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeD =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeB, NodeC));
+
+  auto UpdateNodeA = UpdateGraph.add(EmptyKernel);
+  auto UpdateNodeB = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  auto UpdateNodeC = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  auto UpdateNodeD = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(
+                       UpdateNodeA, UpdateNodeB, UpdateNodeC /* Extra Edge */));
+
+  auto GraphExec = Graph.finalize(experimental::property::graph::updatable{});
+  EXPECT_THROW(GraphExec.update(UpdateGraph), sycl::exception);
+}
+
+TEST_F(WholeGraphUpdateTest, MissingEdges) {
+  // Test that using an update graph with missing edges results in an error.
+
+  auto NodeA = Graph.add(EmptyKernel);
+  auto NodeB =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeC =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeA));
+  auto NodeD =
+      Graph.add(EmptyKernel, exp_ext::property::node::depends_on(NodeB, NodeC));
+
+  auto UpdateNodeA = UpdateGraph.add(EmptyKernel);
+  auto UpdateNodeB = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  auto UpdateNodeC = UpdateGraph.add(
+      EmptyKernel, exp_ext::property::node::depends_on(UpdateNodeA));
+  auto UpdateNodeD = UpdateGraph.add(
+      EmptyKernel,
+      exp_ext::property::node::depends_on(/* Missing Edge */ UpdateNodeB));
+
+  auto GraphExec = Graph.finalize(experimental::property::graph::updatable{});
+  EXPECT_THROW(GraphExec.update(UpdateGraph), sycl::exception);
+}
+
+// FIXME TODO Is this an error or not?
```

Comment:
It might be worth checking that backends don't consider this an error, I have a vague recollection that maybe CUDA does.

Reply:
I tested it locally by copying the unit test to an end to end test with the cuda backend. It didn't throw any error so I'm assuming it works.

Reply:
Added the following wording to the spec:

"Note: Edges can be added in any order without changing the topology."

https://github.com/intel/llvm/pull/13253/files

Reply:
> It might be worth checking that backends don't consider this an error, I have a vague recollection that maybe CUDA does.


Looking in the description of `cudaGraphExecUpdate`

```
cudaGraphExecUpdate sets the result member of resultInfo to cudaGraphExecUpdateErrorTopologyChanged under the following conditions: 
...
* A node in hGraph has a dependency that does not match with the corresponding dependency of the paired node from hGraphExec. ... The dependencies are paired based on edge order and a dependency does not match when the nodes are already paired based on other edges examined in the graph. 

"dependencies are paired based on edge order" does imply that CUDA-Graph is treating this as an error when we're not. I'd suggest we look into doing this as a follow-up task.

Reply:
As discussed offline, I've created internal tickets to implement node & edge ordering checks as a follow-up task.

I'd change this test to `EXPECT_THROW(GraphExec.update(UpdateGraph));`  and then `TEST_F(WholeGraphUpdateTest, DISABLED_WrongOrderEdges` so that gtest skips it until will implement that and it can be re-enabled.